### PR TITLE
Restore 32-bit support in Java Coder

### PR DESCRIPTION
**Pull request by andriydruk**

### CODE REVIEW

File: Sources/JavaCoderConfig.swift [hunk 1]
```diff
@@ -91,10 +91,18 @@ public struct JavaCoderConfig {
                 let context = EncodingError.Context(codingPath: $1, debugDescription: errorDescription)
                 throw EncodingError.invalidValue(value, context)
             }
+            #if arch(x86_64) || arch(arm64)
```
Comment (andriydruk) [inline Sources/JavaCoderConfig.swift]: @zayass can we rely on `arch(x86_64) || arch(arm64)` with our swift compiler

**Comment by ynnadrules** [comment]
@andriydruk The changes you have made here seem to work except when it comes to doubles. Doubles coming back from the swift side are `NaN` on the kotlin side. Any ideas?

Ex: 
```swift
public struct Progress: Codable {
    public let current: Int
    public let total: Int
    public var progress: Double {
         return Double(current) / Double(total)
    }
}
```
```kotlin
@SwiftValue
class Progress private constructor() {
    @get:SwiftGetter
    val percentage: Double
		external get
}

val foo = bar.progress

foo.percentage == Double.NaN // percentage is NaN
```

**Comment by ynnadrules** [comment]
@andriydruk I was going to add a failing test case in `swift-java-codegen` but all you need to do is point that sample package's `swift-java-coder` dependency to this `dev/32-bit-support` branch and run the tests in `DoubleTests.java`. Unrelated note: the params for the actual and expected values in all the `assert*` calls are flipped 😄

**Comment by andriydruk** [comment]
I've just checked DoubleTest everything works fine. I've just added x86 CI, all tests are green https://github.com/readdle/swift-java-codegen/pull/31
Could you add your test that fails on x86?

**Comment by ynnadrules** [comment]
@andriydruk Hey, well I think due to the latest changes you just pushed to this branch, the test are all green for me too! Before, some of the tests in DoubleTests.java from swift-java-codegen were failing when using the 32-bit-support branch of swift-java-coder, like `testZero()` and `testEncode()`. 

So what do you ultimately think the issue was behind doubles becoming NaNs? The `testZero()` scenario was simply bridging `0.0` which (i believe) doesn't go through the encoding/decodable closures that are registered in JavaCoderConfig.swift, but instead, via JavaObjectEncoder, has `javaPrimitive()` called on the `Double`. 

Anyway, thanks so much for providing a fix for this!

**Comment by andriydruk** [comment]
Ok, looks that everything now works fine
I will merge this PR to `dev/kotlin-support` and merge readdle/swift-java-codegen#31 after branch update

**Comment by ynnadrules** [comment]
@andriydruk It unfortunately turns out the issue is still happening. All the tests in DoubleTests.swift that were failing before were fixed with your last change, but in some cases the NaN issue is still happening for us. Here's a branch in our fork of swift-java-codegen with the bug test case: https://github.com/heliumfoot/swift-java-codegen/tree/bug/32-bit-double-nan-bug

The example is a type that represents progress and has a a computed property that returns a percentage. That percentage is a `Double` and when you use that property on the android side it is `NaN`. Maybe the issue is something obvious that I am not seeing?

Here is the test run: https://github.com/heliumfoot/swift-java-codegen/runs/2223011181?check_suite_focus=true

**Comment by andriydruk** [comment]
Hi, @ynnadrules  thank you

We should definitely start from PR next time and not guessing what broken.
Yes, a problem occurs on a 32-bit device. I've just found an issue in the compiler and fixed it https://github.com/readdle/swift-java-codegen/commit/c62a0892a353d3afc6fcab499a914c0677223710

Could I use tests from your fork in our repo?

**Comment by ynnadrules** [comment]
Hey @andriydruk, yes feel free to use the tests. I unfortunately had to revert to the java-based bridging during these two weeks due to our project's timeline and many of our stakeholders having 32-bit based android phones. But hopefully now we have a path forward with the kotlin-based bridging. I'll test it out soon. 

Out of curiosity, is Readdle using the kotlin-based bridging for their released apps already or just internally or on beta releases?

thanks again @andriydruk.

**Comment by andriydruk** [comment]
We still using Java-bridging in Spark and Kotlin-briding in some internal tools.
But we have the plan to migrate all our Spark bridging code to Kotlin in the next few weeks

We really appreciate your help in the stabilization code-gen 🙏

PS: I pushed your tests to https://github.com/readdle/swift-java-codegen/pull/31

**Pull request merged**